src/lnurl.js

```diff
@@ -0,0 +1,116 @@
+    res.status(200).send(
+      (await listLnurlPayEndpoints())
+        .map(lnurlpay => addBech23Lnurl(req, lnurlpay))
+    )))
+
+  app.post('/endpoint', auth, wrap(async (req, res) =>
+    res.status(201).send(
+      addBech23Lnurl(req, await setLnurlPayEndpoint(null, req.body))
+    )))
+
+  app.put('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      addBech23Lnurl(req, await setLnurlPayEndpoint(req.params.id, req.body))
+    )))
+
+  app.delete('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(await delLnurlPayEndpoint(req.params.id))))
+
+  app.get('/endpoint/:id', auth, wrap(async (req, res) =>
+    res.status(200).send(
+      addBech23Lnurl(req, await getLnurlPayEndpoint(req.params.id))
+    )))
+
+  app.get('/endpoint/:id/invoices', auth, wrap(async (req, res) =>
+    res.send(await listInvoicesByLnurlPayEndpoint(req.params.id))))
+
+  // this is the actual endpoint users will hit
+  app.get('/lnurl/:id', wrap(async (req, res) => {
+    const lnurlpay = await getLnurlPayEndpoint(req.params.id)
+
+    res.status(200).send({
+      tag: 'payRequest'
+    , minSendable: lnurlpay.min
+    , maxSendable: lnurlpay.max
+    , metadata: makeMetadata(lnurlpay)
+    , commentAllowed: lnurlpay.comment
+    , callback: `https://${req.hostname}/lnurl/${lnurlpay.id}/callback`
+    })
+  }))
+
+  app.get('/lnurl/:id/callback', wrap(async (req, res) => {
+    const lnurlpay = await getLnurlPayEndpoint(req.params.id)
+
+    if (req.query.amount > lnurlpay.max)
+      return res.send({status: 'ERROR', reason: 'amount too large'})
+    if (req.query.amount < lnurlpay.min)
+      return res.send({status: 'ERROR', reason: 'amount too small'})
+
+    let invoiceMetadata = {...req.query}
+    delete invoiceMetadata.amount
+    delete invoiceMetadata.fromnodes
+    delete invoiceMetadata.nonce
+    invoiceMetadata = {...lnurlpay.metadata, ...invoiceMetadata}
+
+    const invoice = await newInvoice({
+      descriptionHash: require('crypto')
+        .createHash('sha256')
+        .update(makeMetadata(lnurlpay))
+        .digest('hex')
+    , msatoshi: req.query.amount
+    , metadata: invoiceMetadata
+    , webhook: lnurlpay.webhook
+    , lnurlpay_endpoint: lnurlpay.id
+    })
+
+    let successAction
+    if (lnurlpay.success_url) {
+      successAction = {
+        tag: 'url'
+      , url: lnurlpay.success_url
+      , description: lnurlpay.success_text || ''
+      }
+    } else if (lnurlpay.success_value) {
+      // not implemented yet
+    } else if (lnurlpay.success_text) {
+      successAction = {tag: 'message', message: lnurlpay.success_text}
+    }
+
+    res.status(200).send({
+      pr: invoice.payreq
+    , successAction
+    , routes: []
+    , disposable: false
+    })
+  }))
+}
+
+function makeMetadata (lnurlpay) {
+  const text = lnurlpay.text
+
+  const meta = [['text/plain', text]]
+    .concat(lnurlpay.image ? ['image/png;base64', lnurlpay.image] : [])
```

Comment:
Maybe let users add additional custom metadata tags to make this more future-proof? Something like `.concat(lnurlpay.metadata.custom_tags || [])`

Reply:
This "metadata" here is not the same metadata charge uses and stores, it's just a way to transfer a description and an image to the wallet since the invoice uses `description_hash` instead of `description`.

Reply:
Yes, I understand, my suggestion was to support populating the lnurl-pay's metadata array with custom fields provided by the user via charge's metadata. So if new metadata field types are added in the future to the lnurl-pay specs, it will be possible to start using them without requiring additional changes in charge.